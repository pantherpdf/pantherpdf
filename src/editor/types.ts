/**
 * types.ts
 * types used in report editor
 */

import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import type { FunctionComponent, ReactNode, CSSProperties } from 'react';
import type { FormulaHelper } from './compile';
import {
  TReport,
  TData,
  TTransformData,
  ApiEndpoints,
  TDataCompiled,
  TReportCompiled,
} from '../types';

export type TName = string | { [key: string]: string };
export interface TFontAwesomeIcon {
  fontawesome: IconDefinition;
}

// to help construct tests
// to force specific children type
export type ForceChildren<T> =
  | T
  | { [key: string]: unknown; children: ForceChildren<T>[] };

export type ReportForceChildren<T> = TReport & { children: ForceChildren<T>[] };

export interface TSourceData {
  data: unknown;
  errorMsg?: string;
}

export type TDragObj =
  | { type: 'wid'; wid: number[] }
  | { type: 'widget'; widget: TData }
  | { type: 'widgets'; widgets: TData[] };

export interface GeneralProps {
  getOriginalSourceData: () => Promise<unknown>;
  overrideSourceData?: (data: unknown) => void;
  isOverridenSourceData: boolean;
  data: TSourceData;
  api: ApiEndpoints;

  report: TReport;
  setReport: (report: TReport) => Promise<void>;
  deleteReport?: () => void;

  selected: number[] | null;
  setSelected: React.Dispatch<React.SetStateAction<number[] | null>>;

  renderWidget: (child: TData, parents: number[]) => ReactNode;
  renderWidgets: (children: TData[], parents: number[]) => ReactNode;
  dragWidgetStart: (
    e: React.DragEvent<HTMLDivElement>,
    dragObj: TDragObj,
  ) => void;
  dragWidgetEnd: (e: React.DragEvent<HTMLDivElement>) => void;
  drop: (e: React.DragEvent<HTMLDivElement>, dest: number[]) => void;
}

export interface ItemRendeProps extends GeneralProps {
  item: TData;
  setItem: (itm: TData) => void;
  wid: number[];
}

export interface ItemRendeFinalHelper {
  renderItem: (item: TDataCompiled, helper: ItemRendeFinalHelper) => string;
  renderChildren: (
    chs: TDataCompiled[],
    helper: ItemRendeFinalHelper,
  ) => string;
  externalHelpers: { [key: string]: any };
  escapeHtml: (txt: string) => string;
  styleToStringAttribute: (style: CSSProperties) => string;
}

export interface ItemRendeFinalProps extends ItemRendeFinalHelper {
  item: TDataCompiled;
}

export interface CompileHelper {
  formulaHelper: FormulaHelper;
  evalFormula: (txt: string) => Promise<unknown>;
  compileChildren: (
    arr1: TData[],
    helper: CompileHelper,
  ) => Promise<TDataCompiled[]>;
  wid: number[];
  report: TReport;
  reportCompiled: TReportCompiled;
  api: ApiEndpoints;
  externalHelpers: { [key: string]: any };
  variables: { [key: string]: unknown };
}

export interface NewItemProps {
  report: TReport;
}

export interface Widget {
  name: TName;
  icon: TFontAwesomeIcon;
  newItem: (props: NewItemProps) => Promise<TData>;
  compile: (dt: any, helper: CompileHelper) => Promise<TDataCompiled>;
  RenderProperties?: FunctionComponent<ItemRendeProps>;
  Render: FunctionComponent<ItemRendeProps>;
  RenderFinal: (props: ItemRendeFinalProps) => string;
  canAdd?: boolean;
  canDrag?: boolean;
  canSelect?: boolean;
}

export interface TransformRendeProps extends GeneralProps {
  item: TTransformData;
  setItem: (itm: TTransformData) => void;
  index: number;
}

export interface TTransformWidget {
  name: TName;
  newItem: () => Promise<TTransformData>;
  transform: (
    dt: unknown,
    item: TTransformData,
    allowUnsafeJsEval: boolean,
  ) => Promise<unknown>;
  Editor: FunctionComponent<TransformRendeProps>;
}
