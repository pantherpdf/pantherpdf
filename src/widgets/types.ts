/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2023
 * @license MIT
 */

import type { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import type { FunctionComponent } from 'react';
import type { ApiEndpoints, Name, ReportCompiled, Report } from '../types';
import type { FormulaHelper } from '../data/compile';
import type { GeneralProps } from '../editor/types';

/** Building blocks for editor */
export interface Widget {
  /** Unique widget id (type) */
  id: string;

  /** Human friendly name */
  name: Name;

  icon: IconDefinition;

  /** Create new instance of this widget */
  newItem: (props: ItemNewProps) => Promise<Item>;

  /** Compile item */
  compile: (item: any, helper: CompileHelper) => Promise<ItemCompiled>;

  /** Render properties editor */
  RenderProperties?: FunctionComponent<ItemRenderEditorProps>;

  /** Render item inside editor */
  RenderEditor: FunctionComponent<ItemRenderEditorProps>;

  /** Render item for final report */
  RenderPreview: FunctionComponent<ItemRenderPreviewProps>;

  /**
   * Can user add this widget to report? Default true.
   * Useful for helper Widgets which user should not manually manipulate.
   */
  canAdd?: boolean;

  /** Can user drag this widget to report (inside editor)? Default true. */
  canDrag?: boolean;

  /** Can user select Item by clicking ot it? Default true. */
  canSelect?: boolean;
}

/**
 * Instance of a Widget.
 * Reports are made out of multiple items. It contains data like text,
 * font-size, image url, margins etc.
 */
export interface Item {
  [key: string]: unknown;

  /** Id of a Widget */
  type: string;

  /** Subitems */
  children: Item[];
}

export interface ItemCompiled {
  [key: string]: unknown;
  type: string;
}

export interface ItemNewProps {
  report: Report;
}

export interface CompileHelper {
  formulaHelper: FormulaHelper;
  evalFormula: (txt: string) => Promise<unknown>;
  compileChildren: (
    children: Item[],
    helper: CompileHelper,
  ) => Promise<ItemCompiled[]>;

  /** Hierarchical index for current item */
  wid: number[];

  report: Report;
  reportCompiled: ReportCompiled;
  api: ApiEndpoints;
  externalHelpers: { [key: string]: any };
  variables: { [key: string]: unknown };
}

export interface ItemRenderEditorProps extends GeneralProps {
  item: Item;
  setItem: (itm: Item) => void;
  wid: number[];
}

export interface ItemRenderPreviewHelper {
  renderItem: (
    item: ItemCompiled,
    helper: ItemRenderPreviewHelper,
  ) => React.ReactNode;
  renderChildren: (
    chs: ItemCompiled[],
    helper: ItemRenderPreviewHelper,
  ) => React.ReactNode[];
  externalHelpers: { [key: string]: any };
}

export interface ItemRenderPreviewProps extends ItemRenderPreviewHelper {
  item: ItemCompiled;
}
