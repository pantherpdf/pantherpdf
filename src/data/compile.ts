/**
 * @file Compile sourceData and report - evaluate all formulas
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

import type {
  Report,
  ApiEndpoints,
  ReportCompiled,
  ReportProperties,
  ReportPropertiesCompiled,
  FormulaObject,
} from '../types';
import formulaEvaluate from '../formula/formula';
import { getWidget } from '../widgets/allWidgets';
import type { CompileHelper, Widget, WidgetItem } from '../widgets/types';
import { PropertyFontExtractStyle } from '../widgets/PropertyFont';

type TOvrr = [string, unknown];

export class FormulaHelper {
  overrides: TOvrr[];

  constructor() {
    this.overrides = [];
  }

  push(name: string, valueOrCb: unknown) {
    this.overrides.push([name, valueOrCb]);
  }

  pop() {
    this.overrides.pop();
  }

  async getVar(name: string): Promise<unknown> {
    for (let i = this.overrides.length - 1; i >= 0; --i) {
      if (this.overrides[i][0] === name) {
        let val = this.overrides[i][1];
        if (typeof val == 'function') {
          val = await val(name);
        }
        return val;
      }
    }
    return undefined;
  }
}

export default async function compile(
  report: Report,
  data: unknown,
  widgets: Widget[],
  api: ApiEndpoints,
  externalHelpers: { [key: string]: unknown } = {},
): Promise<ReportCompiled> {
  // make a copy, to support changing
  report = JSON.parse(JSON.stringify(report));

  const formulaHelper = new FormulaHelper();
  formulaHelper.push('data', data);
  formulaHelper.push('report', report);
  const evalFormulaWrapper = async (val: FormulaObject): Promise<unknown> => {
    return formulaEvaluate(val.formula, formulaHelper);
  };

  // custom variables
  const vars: { [key: string]: unknown } = {};
  function getVarValue(varName: string): unknown {
    return vars[varName];
  }
  for (const v of report.variables) {
    vars[v.name] = evalFormulaWrapper(v.value);
    formulaHelper.push(v.name, getVarValue);
  }

  const propsCompiled = await compilePrprt(
    report.properties,
    evalFormulaWrapper,
  );

  const helper: CompileHelper = {
    wid: [],
    report: report,
    propertiesCompiled: propsCompiled,
    formulaHelper,
    api,
    variables: vars,
    externalHelpers,

    evalFormula: evalFormulaWrapper,

    compileChildren: async (children: WidgetItem[], helper: CompileHelper) => {
      // dont use promise.all() because order of execution matters and some async operation could change it
      const arr2 = [];
      for (let i = 0; i < children.length; ++i) {
        const ch = children[i];
        const helper2 = { ...helper, wid: [...helper.wid, i] };
        const dt = await getWidget(widgets, ch.type).compile(ch, helper2);
        arr2.push(dt);
      }
      return arr2;
    },
  };

  const widgetsCompiled = await helper.compileChildren(report.widgets, helper);

  formulaHelper.pop();
  formulaHelper.pop();
  report.variables.map(() => formulaHelper.pop());
  if (formulaHelper.overrides.length !== 0) {
    throw new Error('helper has overrides still left inside');
  }

  return {
    widgets: widgetsCompiled,
    properties: propsCompiled,
  };
}

async function compilePrprt(
  prprt: ReportProperties,
  evalFormulaWrapper: (formula: FormulaObject) => Promise<unknown>,
): Promise<ReportPropertiesCompiled> {
  const obj: ReportPropertiesCompiled = {
    // make deep copy because widgets can change properties
    ...JSON.parse(JSON.stringify(prprt)),
    fileName: undefined,
    fontsUsed: [],
    globalCss: '',
  };
  if (prprt.fileName !== undefined) {
    const res = await evalFormulaWrapper(prprt.fileName);
    if (typeof res !== 'string') {
      throw new Error(
        `Filename should be a string, but received ${typeof res}`,
      );
    }
    obj.fileName = res;
  }
  if (prprt.font) {
    const style = PropertyFontExtractStyle(prprt.font);
    if (style) {
      obj.fontsUsed.push(style);
    }
  }
  return obj;
}
