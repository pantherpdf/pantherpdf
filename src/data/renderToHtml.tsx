/**
 * @file Compiled report -> html
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

import React, { CSSProperties } from 'react';
import { renderToString } from 'react-dom/server';
import { defaultReportCss } from '../types';
import type { ApiEndpoints, ReportCompiled } from '../types';
import type { WidgetPreviewPropsBase, Widget } from '../widgets/types';
import { getWidget } from '../widgets/allWidgets';
import { PropertyFontGenCss } from '../widgets/PropertyFont';

export function renderBody(
  report: ReportCompiled,
  widgets: Widget[],
  externalHelpers: { [key: string]: unknown } = {},
): React.ReactElement {
  const helper: WidgetPreviewPropsBase = {
    renderItem: (item, helper) => {
      const w = getWidget(widgets, item.type);
      return w.Preview({ ...helper, item });
    },
    renderChildren: (chs, helper) => {
      if (!Array.isArray(chs)) {
        throw new Error('Bad childs array');
      }
      return chs.map((item, idx) => {
        const w = getWidget(widgets, item.type);
        return (
          <React.Fragment key={idx}>
            {w.Preview({ ...helper, item })}
          </React.Fragment>
        );
      });
    },
    externalHelpers,
  };

  const children = helper.renderChildren(report.widgets, helper);

  // prepare css
  const cssObj: CSSProperties = {
    ...defaultReportCss,
    ...PropertyFontGenCss(report.properties.font || {}),
  };
  return <body style={cssObj}>{children}</body>;
}

export default function renderToHtml(
  report: ReportCompiled,
  widgets: Widget[],
  api: ApiEndpoints,
  externalHelpers: { [key: string]: unknown } = {},
): string {
  // render content
  const bodyElement = renderBody(report, widgets, externalHelpers);
  const bodyTxt = renderToString(bodyElement);

  const fontUrl = api.fonts?.getCssUrls(report.properties.fontsUsed) || [];
  const fontHtml = fontUrl
    .map(url => `<link rel="stylesheet" href="${url}"></link>`)
    .join('\n');

  const html = `<!DOCTYPE html>
<html lang="en-US">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title>Report</title>
<style>
/* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/
html, body, div, span, applet, object, iframe,
h1, h2, h3, h4, h5, h6, p, blockquote, pre,
a, abbr, acronym, address, big, cite, code,
del, dfn, em, img, ins, kbd, q, s, samp,
small, strike, strong, sub, sup, tt, var,
b, u, i, center,
dl, dt, dd, ol, ul, li,
fieldset, form, label, legend,
table, caption, tbody, tfoot, thead, tr, th, td,
article, aside, canvas, details, embed, 
figure, figcaption, footer, header, hgroup, 
menu, nav, output, ruby, section, summary,
time, mark, audio, video {
	margin: 0;
	padding: 0;
	border: 0;
	font-size: 100%;
	font: inherit;
	vertical-align: baseline;
}
/* HTML5 display-role reset for older browsers */
article, aside, details, figcaption, figure, 
footer, header, hgroup, menu, nav, section {
	display: block;
}
body {
	line-height: 1.25;
}
ol, ul {
	list-style: none;
}
blockquote, q {
	quotes: none;
}
blockquote:before, blockquote:after,
q:before, q:after {
	content: '';
	content: none;
}
table {
	border-collapse: collapse;
	border-spacing: 0;
}
/* modifications */
b, strong {
	font-weight: bold;
}
cite, em, i {
	font-style: italic;
}
big {
	font-size: larger;
}
small {
	font-size: smaller;
}
mark {
	background-color: yellow;
}
sub {
    vertical-align: sub;
    font-size: smaller;
}
sup {
    vertical-align: super;
    font-size: smaller;
}
</style>
<style>
* {
	-webkit-print-color-adjust: exact;
	print-color-adjust: exact;
}
</style>
<style>
${report.properties.globalCss}
</style>
${fontHtml}
</head>
${bodyTxt}
</html>
`;
  return html;
}
