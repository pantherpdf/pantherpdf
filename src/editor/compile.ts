/**
 * @file Compile sourceData and report - evaluate all formulas
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021
 * @license MIT
 */

import { Report, Item, ApiEndpoints } from '../types';
import type { ReportCompiled } from '../types';
import FormulaEvaluate from '../formula/formula';
import { getWidget } from '../widgets/allWidgets';
import { CompileHelper, Widget } from './types';
import { PropertyFontExtractStyle } from '../widgets/PropertyFont';

type TOvrr = [string, unknown];

export class FormulaHelper {
  overrides: TOvrr[];

  constructor() {
    this.overrides = [];
  }

  push(name: string, valueOrCb: unknown) {
    this.overrides.push([name, valueOrCb]);
  }

  pop() {
    this.overrides.pop();
  }

  async getVar(name: string): Promise<unknown> {
    for (let i = this.overrides.length - 1; i >= 0; --i) {
      if (this.overrides[i][0] === name) {
        let val = this.overrides[i][1];
        if (typeof val == 'function') {
          val = await val(name);
        }
        return val;
      }
    }
    return undefined;
  }
}

export default async function compile(
  report: Report,
  data: unknown,
  widgets: Widget[],
  api: ApiEndpoints,
  externalHelpers: { [key: string]: any } = {},
): Promise<ReportCompiled> {
  // make a copy, to support changing
  report = JSON.parse(JSON.stringify(report));

  const formulaHelper = new FormulaHelper();
  const getVar = formulaHelper.getVar.bind(formulaHelper);
  formulaHelper.push('data', data);
  formulaHelper.push('report', report);

  // custom variables
  const vars: { [key: string]: unknown } = {};
  function getVarValue(varName: string): unknown {
    return vars[varName];
  }
  for (const v of report.variables) {
    vars[v.name] = await FormulaEvaluate(v.formula, { getVar });
    formulaHelper.push(v.name, getVarValue);
  }

  let dt2: ReportCompiled = {
    ...report,
    children: [],
    fontsUsed: [],
    globalCss: '',
  };
  dt2 = JSON.parse(JSON.stringify(dt2));
  if (dt2.properties.font) {
    const style = PropertyFontExtractStyle(dt2.properties.font);
    if (style) {
      dt2.fontsUsed.push(style);
    }
  }

  if (dt2.properties.fileName) {
    const res = await FormulaEvaluate(dt2.properties.fileName, { getVar });
    if (typeof res !== 'string') {
      throw new Error(
        `Filename should be a string, but received ${typeof res}`,
      );
    }
    dt2.properties.fileName = res;
  }

  const helper: CompileHelper = {
    wid: [],
    report: report,
    reportCompiled: dt2,
    formulaHelper,
    api,
    variables: vars,
    externalHelpers,

    evalFormula: async (txt: string) => {
      return FormulaEvaluate(txt, { getVar });
    },

    compileChildren: async (children: Item[], helper: CompileHelper) => {
      // dont use promise.all() because order of execution matters and some async operation could change it
      const arr2 = [];
      for (let i = 0; i < children.length; ++i) {
        const ch = children[i];
        const helper2 = { ...helper, wid: [...helper.wid, i] };
        const dt = await getWidget(widgets, ch.type).compile(ch, helper2);
        arr2.push(dt);
      }
      return arr2;
    },
  };

  dt2.children = await helper.compileChildren(report.children, helper);

  formulaHelper.pop();
  formulaHelper.pop();
  report.variables.map(() => formulaHelper.pop());
  if (formulaHelper.overrides.length !== 0) {
    throw new Error('helper has overrides still left inside');
  }
  return dt2;
}
