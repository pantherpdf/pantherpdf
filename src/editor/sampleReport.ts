/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2023
 * @license MIT
 */

import type { Report } from '../types';

export const sampleReport: Report = {
  name: 'Report 1',
  widgets: [],
  transforms: [],
  properties: {},
  variables: [],
};
