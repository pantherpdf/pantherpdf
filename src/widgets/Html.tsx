/**
 * Html.tsx
 * Render html
 */


import React from 'react'
import { TData, TDataCompiled } from '../types'
import type { Widget } from '../editor/types'
import { faCode } from '@fortawesome/free-solid-svg-icons'
import BoxName from './BoxName'
import InputApplyOnEnter from './InputApplyOnEnter'
import Trans from '../translation'


export interface HtmlData extends TData {
	type: 'Html',
	source: string,
}

export interface HtmlCompiled extends TDataCompiled {
	type: 'Html',
	data: string,
}


export const Html: Widget = {
	name: {en: 'Html', sl: 'Html'},
	icon: {fontawesome: faCode},

	newItem: async (): Promise<HtmlData> => {
		return {
			type: 'Html',
			children: [],
			source: '',
		}
	},

	compile: async (dt: HtmlData, helpers): Promise<HtmlCompiled> => {
		const str2 = await helpers.evalFormula(dt.source)
		const str = (str2 !== undefined && str2 !== null && str2 !== false) ? String(str2) : ''
		return {
			type: dt.type,
			children: [],
			data: str,
		}
	},

	Render: function(props) {
		const item = props.item as HtmlData
		return <BoxName {...props} name={Html.name}>
			<div className='font-monospace'>
				{item.source}
			</div>
		</BoxName>
	},

	RenderFinal: function(props) {
		const item = props.item as HtmlCompiled
		return <div
			dangerouslySetInnerHTML={{__html: item.data}}
		/>
	},

	RenderProperties: function(props) {
		const item = props.item as HtmlData
		return <>
			<label htmlFor='Html-source' className='d-block'>
				{Trans('source data')}
			</label>
			<div className="input-group mb-3">
				<span className="input-group-text fst-italic">Æ’</span>
				<InputApplyOnEnter
					id='Html-source'
					value={item.source}
					onChange={val=>props.setItem({...item, source: val})}
				/>
			</div>
		</>
	},
}
