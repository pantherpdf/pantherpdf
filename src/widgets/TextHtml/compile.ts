import { listOfAdjusts } from '../formulaAdjust';
import { CompileHelper, WidgetItem } from '../types';
import type { TextHtmlCompiled, TextHtmlData } from './TextHtml';

export default async function compile(
  item: WidgetItem,
  helper: CompileHelper,
): Promise<TextHtmlCompiled> {
  const dt = item as TextHtmlData;
  // combine parts
  let html = '';
  for (const part of dt.value) {
    const type = part.type;
    switch (type) {
      case 'html': {
        html += part.value;
        break;
      }
      case 'formula': {
        let value = await helper.evalFormula(part);
        if (part.adjust && part.adjust.length > 0) {
          const adjustObj = listOfAdjusts.find(x => x.id === part.adjust);
          if (!adjustObj) {
            throw new Error(`Unknown adjust ${part.adjust}`);
          }
          value = adjustObj.func(value, []);
        }
        html += String(value);
        break;
      }
      default: {
        const exhaustiveCheck: never = type;
        throw new Error(`Unknown data type: ${exhaustiveCheck}`);
      }
    }
  }
  return {
    type: dt.type,
    value: html,
    font: dt.font,
  };
}
