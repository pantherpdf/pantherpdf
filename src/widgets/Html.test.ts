/**
 * @jest-environment node
 */

import { HtmlData, HtmlCompiled } from './Html';
import compile, { compileComponent } from '../editor/compile';
import { renderToHtmlContent } from '../editor/renderToHtml';
import { sampleReport } from '../editor/sampleReport';
import { ReportForceChildren } from '../editor/types';

test('Html compile', async () => {
  const dt: HtmlData = { type: 'Html', source: 'data.desc', children: [] };

  const data = { desc: '<p>Hello</p>' };
  const p2 = await compileComponent(dt, data);
  expect(p2).toBeTruthy();
  expect(p2.type).toBe('Html');
  const p = p2 as HtmlCompiled;
  expect(p.data).toBe('<p>Hello</p>');
});

test('html render', async () => {
  const report: ReportForceChildren<HtmlData> = {
    ...sampleReport,
    children: [{ type: 'Html', source: 'data.desc', children: [] }],
  };

  const data = { desc: '<p>Hello</p>' };
  const compiled = await compile(report, data);
  const html = renderToHtmlContent(compiled);
  expect(html).toMatchSnapshot();
});

test('Html source==null should be empty string', async () => {
  const dt: HtmlData = { type: 'Html', source: 'data.dt', children: [] };
  const p2 = await compileComponent(dt, { dt: null });
  expect(p2.data).toBe('');
});

test('Html source==false should be empty string', async () => {
  const dt: HtmlData = { type: 'Html', source: 'data.dt', children: [] };
  const p2 = await compileComponent(dt, { dt: false });
  expect(p2.data).toBe('');
});

test('Html source==undefined should be empty string', async () => {
  const dt: HtmlData = { type: 'Html', source: 'data.dt', children: [] };
  const p2 = await compileComponent(dt, { dt: undefined });
  expect(p2.data).toBe('');
});

test('Html source==0 should be "0"', async () => {
  const dt: HtmlData = { type: 'Html', source: 'data.dt', children: [] };
  const p2 = await compileComponent(dt, { dt: 0 });
  expect(p2.data).toBe('0');
});
