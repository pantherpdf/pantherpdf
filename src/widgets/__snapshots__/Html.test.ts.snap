// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html render 1`] = `
<div>
  <p>
    Hello
  </p>
</div>
`;
